### PR TITLE
Fix VarAutoEncoder reparameterize returning mu+std at inference (#8413)

### LINKED ISSUE #8413: Error in VarAutoEncoder class
I'm wondering if there isn't a bug in the implementation of the VarAutoEncoder class.
For the reparameterisation trick (lines 144-150) https://github.com/Project-MONAI/MONAI/blob/46a5272196a6c2590ca2589029eed8e4d56ff008/monai/networks/nets/varautoencoder.py#L28-L155

def reparameterize(self, mu: torch.Tensor, logvar: torch.Tensor) -> torch.Tensor:
    std = torch.exp(0.5 * logvar)

    if self.training:  # multiply random noise with std only during training
        std = torch.randn_like(std).mul(std)

    return std.add_(mu)

The code is correct for training. At test/inference time, we don't multiple std by a random epsilon, which is correct. However, the function still returns mu+std, which doesn't make sense to me. It should be returning mu only at test time, it makes no sense to take mu+std. Can someone check if I'm correct?

### PR DESCRIPTION
**Pull request by Shizoqua**
### Description
Fixes #8413. `VarAutoEncoder.reparameterize` computed `std = exp(0.5*logvar)`, added the random noise term only when training, and then returned `std.add_(mu)`. At inference (`eval()`), `std` is still the standard deviation, so the method returned `mu + std` instead of the posterior mean `mu`. The reparameterization trick's stochastic term should apply during training only; at inference the latent code should be `mu`.

The fix returns `mu` directly when not training, and computes `mu + eps * std` out-of-place otherwise (also avoids the in-place `add_`).

`VarFullyConnectedNet.reparameterize` (`monai/networks/nets/fullyconnectednet.py`) contained the identical bug, so it is fixed the same way in this PR.

### Types of changes
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.

### Testing
Added regression tests asserting that in eval mode the returned latent equals `mu` and is deterministic, while training stays stochastic, for both `VarAutoEncoder` and `VarFullyConnectedNet`.

```
python -m unittest tests.networks.nets.test_varautoencoder tests.networks.nets.test_fullyconnectednet
# Ran 12 tests ... OK
```

### CODE REVIEW

**Comment by Shizoqua** [comment]
@ericspod Since #8906 (Windows docs) was merged, this one's still waiting. Fixes #8413 — VarAutoEncoder returning mu+std at inference during eval. Rebased, CI all green, issue still open. Would appreciate a review when you get a chance.

**Comment by Shizoqua** [comment]
@ericspod — This has been ready for review for 27 days. Could you take a look when you have a chance?

**Comment by Shizoqua** [comment]
Rebased onto current \dev\ and addressed the review feedback — added Google-style docstrings to the \
eparameterize\ methods in \ullyconnectednet.py\ and \arautoencoder.py\, and converted the test-method inline comments to proper docstrings in both test files.

**Review by vikashg — DISMISSED** [review]
PR #8921 (by rtmalikian) is the same fix but without tests. Closing #8921 in favor of this one. Thanks @Shizoqua

**Comment by ericspod** [comment]
Hi @Shizoqua @vikashg I'm not sure this is correct in fact, and is a breaking change that isn't accounted for. 

When doing inference with a trained VAE, a sample can be generated by feeding a random variable through the decoder to produce a sample. Here though the variable is from encoding inputs, and the goal is to reconstruct the input. Supposedly adding the standard deviation as in the original formulation produces less blurry results.

Either way, this PR represents a breaking change that would impact other users. For those that want the original behaviour, this should be left as the default so that backward compatibility is maintained, then both classes should have a new constructor argument added to enable the new behaviour. 

Thoughts on this? I really think this should be added before merging, thanks.

**Comment by vikashg** [comment]
@ericspod Let me have a look again... today thanks

**Comment by Shizoqua** [comment]
@ericspod Thanks for the detailed review — you're right that this is a breaking change. I've updated the PR to preserve backward compatibility:

- **Default behaviour is unchanged**: at inference, both classes still return `mu + std` (as before the PR).
- A new `use_mean_at_inference=True` constructor argument opts into returning the posterior mean at eval, which is the behaviour that issue #8413 asked for.
- Added to both `VarAutoEncoder` and `VarFullyConnectedNet`, with regression tests covering the opt-in behaviour and the preserved default.

Commit a0b471d06. Happy to adjust the parameter name if you'd prefer something else.

**Comment by vikashg** [comment]
Thanks @Shizoqua. I will let @ericspod look at this. I think he is more fluent with this than I am. Thank you

File: monai/networks/nets/fullyconnectednet.py [hunk 1]
```diff
@@ -172,12 +176,27 @@ def decode_forward(self, z: torch.Tensor, use_sigmoid: bool = True) -> torch.Ten
...
+        Returns:
+            Sampled latent code, same shape as ``mu``.
+        """
+        if not self.training and self.use_mean_at_inference:
+            # At inference the latent code is the posterior mean; the random
+            # term is only added during training (the reparameterization trick).
+            return mu
+        std = torch.exp(0.5 * logvar)
+        eps = torch.randn_like(std) if self.training else torch.ones_like(std)
+        return mu + eps * std
```
Comment (ericspod) [inline monai/networks/nets/fullyconnectednet.py]:
```suggestion
        std = logvar.mul(0.5).exp_()
        if self.training:
            std = torch.randn_like(std).mul_(std)
        return std.add_(mu)
```
I'd like to make a change back to something like the original to avoid making a large `ones_like` tensor just to do a no-op multiplication, and to use inline operators to avoid further copies.

File: monai/networks/nets/varautoencoder.py [hunk 2]
```diff
@@ -142,12 +146,27 @@ def decode_forward(self, z: torch.Tensor, use_sigmoid: bool = True) -> torch.Ten
...
+            # term is only added during training (the reparameterization trick).
+            return mu
         std = torch.exp(0.5 * logvar)
-
-        if self.training:  # multiply random noise with std only during training
-            std = torch.randn_like(std).mul(std)
-
-        return std.add_(mu)
+        eps = torch.randn_like(std) if self.training else torch.ones_like(std)
+        return mu + eps * std
```
Comment (ericspod) [inline monai/networks/nets/varautoencoder.py]: Same change here as the above to avoid `ones_like` and use inline ops where possible.

**Review by ericspod — CHANGES_REQUESTED** [review]
Thanks @Shizoqua I had some minor change for the formulation to improve memory efficiency, otherwise it looks fine to me. Please have a look at this and the formatting issue but we should be good to go then.

**Comment by Shizoqua** [comment]
@ericspod Thanks for the review. I have cleaned up the branch:

- Rebased onto the latest dev.
- Removed the stray monai/utils/safeeval.py and tests/utils/test_safe_eval.py files that had accidentally leaked into this branch.
- Kept the memory-efficiency tweak: at inference with the default use_mean_at_inference=False, the code now returns mu + std directly instead of allocating torch.ones_like(std).
- Ran black on the two test files to fix the formatting failure.

CI is re-running now. Should be good for another look.

**Comment by Shizoqua** [comment]
Hi @ericspod, thanks for the feedback. I've updated both `VarAutoEncoder` and `VarFullyConnectedNet` to use the more memory-efficient formulation:
- training: `std = torch.randn_like(std).mul_(std); return std.add_(mu)`
- inference with `use_mean_at_inference=True`: returns `mu` directly
- default inference: still returns `mu + std` for backward compatibility

I also rebased onto the latest `upstream/dev` and the diff is now clean (no stray files). Could you take another look when you have a moment?

**Comment by Shizoqua** [comment]
Hi @ericspod, I have rebased this onto the latest upstream/dev (now

**Review by ericspod — APPROVED** [review]
Hi @Shizoqua it looks good now, thanks.

**Comment by ericspod** [comment]
Thanks @vikashg for help on this one as well!
